### PR TITLE
ENH added crossval class

### PR DESCRIPTION
**Pull request by QB3**

### CODE REVIEW

**Comment by QB3** [comment]
Ready to go for me

File: examples/plot_grad_search_cv.py [hunk 1]
```diff
@@ -20,10 +20,11 @@
 from sklearn.datasets import make_regression
 from sklearn.linear_model import LassoCV
 from sparse_ho.models import Lasso
-from sparse_ho.criterion import CV
+from sparse_ho.criterion import CV, CrossVal
 from sparse_ho.implicit_forward import ImplicitForward
 from sparse_ho.utils import Monitor
-from sparse_ho.grad_search_CV import grad_search_CV
+from sparse_ho.ho import grad_search
+# from sparse_ho.grad_search_CV import grad_search_CV
```
Comment (agramfort) [inline examples/plot_grad_search_cv.py:27]: cleanup

File: examples/plot_grad_search_cv.py [hunk 2]
```diff
@@ -78,27 +81,34 @@
 t0 = time.time()
 Model = Lasso
 Criterion = CV
-Algo = ImplicitForward
+# Algo = ImplicitForward
+# Algo = ImplicitForward
```
Comment (agramfort) [inline examples/plot_grad_search_cv.py:85]: cleanup

File: examples/plot_sparse_log_reg.py [hunk 3]
```diff
@@ -27,7 +27,12 @@
 from sparse_ho.forward import Forward
 from sparse_ho.grid_search import grid_search
 from sparse_ho.datasets.real import get_rcv1
-from sparse_ho.datasets.real import get_real_sim
+
+from sklearn.datasets import make_classification
+# from sklearn.datasets import make_regression
```
Comment (agramfort) [inline examples/plot_sparse_log_reg.py:32]: cleanup

File: sparse_ho/criterion.py [hunk 4]
```diff
@@ -1,15 +1,33 @@
...
+from sklearn.model_selection import train_test_split
+
 from sparse_ho.utils import sigma, smooth_hinge
 from sparse_ho.utils import derivative_smooth_hinge
 from sparse_ho.forward import get_beta_jac_iterdiff
-from scipy.sparse import issparse
 
 
 class CV():
+    """Held out loss (we should change the name CV here).
```
Comment (agramfort) [inline sparse_ho/criterion.py:12]: attributes are not documented.

File: sparse_ho/criterion.py [hunk 5]
```diff
@@ -81,7 +99,18 @@ def get_val_grad(
 
 class Logistic():
```
Comment (agramfort) [inline sparse_ho/criterion.py:100]: docstring missing with attributes

File: sparse_ho/criterion.py [hunk 6]
```diff
@@ -145,7 +174,18 @@ def get_val_grad(
 
 class SmoothedHinge():
```
Comment (agramfort) [inline sparse_ho/criterion.py:175]: docstring missing

File: sparse_ho/criterion.py [hunk 7]
```diff
@@ -306,3 +346,80 @@ def get_val_grad(
             grad = None
 
         return val, grad
+
+
+class CrossVal():
```
Comment (agramfort) [inline sparse_ho/criterion.py:351]: class docstring missing

Comment (agramfort) [reply sparse_ho/criterion.py:351]: see on this rendered page https://qb3.github.io/sparse-ho/api.html that all classes and function docstrings are properly formated

File: sparse_ho/criterion.py [hunk 8]
```diff
@@ -306,3 +346,80 @@ def get_val_grad(
...
+            Maximal number of iteration for the soa solver
+        """
+        self.X = X
+        self.y = y
+        self.dict_crits = {}
+        self.val_test = None
+        self.rmse = None
+
+        # init dict of models
+        if isinstance(cv, int):
```
Comment (agramfort) [inline sparse_ho/criterion.py:373]: cv=int defaults to KFold(int) so this branching is not correct. What do you here is a ShuffleSplit in sklearn

File: sparse_ho/criterion.py [hunk 9]
```diff
@@ -306,3 +346,80 @@ def get_val_grad(
...
+            for i, (train, val) in enumerate(cv.split(X)):
+                X_train = X[train, :]
+                y_train = y[train]
+                X_val = X[val, :]
+                y_val = y[val]
+
+                if issparse(X_train):
+                    X_train = X_train.tocsc().copy()
+                if issparse(X_val):
+                    X_val = X_val.tocsc().copy()
```
Comment (agramfort) [inline sparse_ho/criterion.py:400]: why the copy?

File: sparse_ho/implicit_forward.py [hunk 10]
```diff
@@ -5,10 +5,10 @@
 
 class ImplicitForward():
```
Comment (agramfort) [inline sparse_ho/implicit_forward.py:6]: docstring missing

File: sparse_ho/criterion.py (hunk 8 above)
Comment (QB3) [reply sparse_ho/criterion.py:409]: you mean that if cv is an int I should do this :
`kf = KFold(n_splits=5, shuffle=True, random_state=42)`
